### PR TITLE
Avoid false-positive matches on intermediate objects in `ecs@mappings`

### LINKED ISSUE #102794: `ecs@mappings` template breaks dynamic mapping of intermediate object fields
### Elasticsearch Version

083b59122b3cdf5168d2d2351e53fca99009baf5 (8.12-SNAPSHOT)

### Installed Plugins

_No response_

### Java Version

_bundled_

### OS Version

Linux

### Problem Description

The `ecs@mappings` component template's dynamic templates unintentionally match dynamically mapped object fields, attempting to map them as non-object field types, breaking ingestion of that object's child properties.

For example, the dynamic template [`ecs_usage_scaled_float`](https://github.com/elastic/elasticsearch/blob/824d06c8cf7fd4811a0ff42e3554bc70dbfb7e0a/x-pack/plugin/core/template-resources/src/main/resources/ecs%40mappings.json#L145-L153) will match the "usage" object when dynamically mapping `a.usage.b: 123`.

### Steps to Reproduce

```
PUT /_index_template/xyz
{
  "index_patterns": ["xyz"],
  "composed_of": ["ecs@mappings"],
  "template": {
    "mappings": {
      "dynamic": true
    }
  }
}

POST /xyz/_doc
{
  "a.usage.b": 123
}
```

```
{
  "error": {
    "root_cause": [
      {
        "type": "illegal_argument_exception",
        "reason": "Malformed content, found extra data after parsing: END_OBJECT"
      }
    ],
    "type": "illegal_argument_exception",
    "reason": "Malformed content, found extra data after parsing: END_OBJECT"
  },
  "status": 400
}
```

(Also that error message is completely unhelpful.)

### Logs (if relevant)

_No response_

### PR DESCRIPTION
**Pull request by felixbarny**
Fixes #102794

### CODE REVIEW

**Review by axw — APPROVED** [review]
LGTM, thanks! I forgot to come back after my PR merged :sweat_smile:

File: x-pack/plugin/core/template-resources/src/main/resources/ecs@mappings.json [hunk 1]
```diff
@@ -149,27 +156,7 @@
...
-        },
-        {
-          "ecs_usage_long_scaled_float": {
-            "mapping": {
-              "type": "scaled_float",
-              "scaling_factor": 1000
-            },
-            "path_match": "*.usage",
-            "match_mapping_type": "string"
+            "match_mapping_type": ["double", "long", "string"]
```
Comment (ruflin) [inline x-pack/plugin/core/template-resources/src/main/resources/ecs@mappings.json:159]: Nice simplification

File: x-pack/plugin/core/template-resources/src/main/resources/ecs@mappings.json [hunk 2]
```diff
@@ -180,7 +167,8 @@
             "path_match": [
               "location",
               "*.location"
-            ]
+            ],
+            "match_mapping_type": ["object", "string"]
```
Comment (ruflin) [inline x-pack/plugin/core/template-resources/src/main/resources/ecs@mappings.json:171]: Is this needed?

Comment (felixbarny) [reply x-pack/plugin/core/template-resources/src/main/resources/ecs@mappings.json:171]: Let me maybe remove that, as there are more ways to specify a [`geo_point`](https://www.elastic.co/guide/en/elasticsearch/reference/current/geo-point.html) value, that wouldn't covered by this, for example `"location": [ -71.34, 41.12 ]`.

But as we need to include objects in the `match_mapping_type`, it risks false-positive matches for things like `"foo.location.bar":  "foo"`. But `subobjects: false` does come to the rescue here.

**Review by ruflin — APPROVED** [review]

**Pull request merged**